sympy/series/sequences.py

```diff
@@ -711,6 +713,160 @@ def coeff_mul(self, coeff):
+        equal to. For example, if :code:`a(n) = f(a(n - 1), ..., a(n - d))`,
+        then the expression should be :code:`f(a(n - 1), ..., a(n - d))`.
+
+    start : start value of sequence (inclusive)
+
+    stop : stop value of sequence (inclusive)
+
+    Examples
+    ========
+
+    >>> from sympy import Function, symbols
+    >>> from sympy.series.sequences import RecursiveSeq
+
+    >>> y = Function("y")
+    >>> n = symbols("n")
+    >>> fib = RecursiveSeq((0, 1), y(n), y(n - 1) + y(n - 2))
+
+    For value at a particular point
+
+    >>> fib.coeff(3)
+    2
+
+    supports slicing
+
+    >>> fib[:6]
+    [0, 1, 1, 2, 3, 5]
+
+    inspect recurrence
+
+    >>> fib.recurrence
+    Eq(y(n), y(n - 2) + y(n - 1))
+
+    automatically determine degree
+
+    >>> fib.degree
+    2
+
+    See Also
+    ========
+
+    sympy.series.sequences.SeqFormula
+
+    """
+
+    def __new__(cls, initial, y, recurrence, start=0, stop=oo):
+        n = y.args[0]
+        y = y.func
+        k = Wild("k", exclude=(n,))
+        degree = -oo
+
+        # Find all applications of y in the recurrence and check that:
+        #   1. The function y is only being used with a single argument; and
+        #   2. All arguments are n + k for constant negative integers k.
+
+        prev_ys = recurrence.find(y)
+        for prev_y in prev_ys:
+            if len(prev_y.args) != 1:
+                raise TypeError("Recurrence should be in a single variable")
+
+            shift = prev_y.args[0].match(n + k)[k]
+            if not (shift.is_constant() and shift.is_integer and shift < 0):
+                raise TypeError("Recurrence should have constant,"
+                                " negative, integer shifts"
+                                " (found {})".format(prev_y))
+
+            if -shift > degree:
+                degree = -shift
+
+        if len(initial) != degree:
+            raise ValueError("Number of initial terms must equal degree")
+
+        degree = S(degree)
+        start = sympify(start)
+        stop = sympify(stop)
+
+        initial = Tuple(*(sympify(x) for x in initial))
+
+        seq = Basic.__new__(cls, initial, y(n), recurrence, start, stop)
+
+        seq.cache = {y(start + k): init for k, init in enumerate(initial)}
+        seq._start = start
+        seq._stop = stop
+        seq.degree = degree
+        seq.y = y
+        seq.n = n
+        seq._recurrence = recurrence
+
+        return seq
+
+    @property
+    def start(self):
+        """The starting point of the sequence. This point is included"""
+        return self._start
+
+    @property
+    def stop(self):
+        """The ending point of the sequence. This point is included"""
+        return self._stop
+
+    @property
+    def interval(self):
+        """Interval on which sequence is defined."""
+        return (self._start, self._stop)
+
+    def _eval_coeff(self, index):
+        if index < len(self.cache):
```

Comment:
Should this test depend on `start`?

Reply:
Yes, it should be `index - start < len(self.cache)`, or just `y(index) in self.cache`. I'll fix this in an upcoming commit.